### PR TITLE
PHPC-1211: Int64 class to wrap 64-bit integers on 32-bit platforms

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-1211

### CODE REVIEW

**Comment by jmikola** [comment]
Test results from a xenial32 Vagrant box:

```
vagrant@ubuntu-xenial:~/mongo-php-driver$ make test TESTS=tests/bson/bson-int64*phpt TEST_PHP_ARGS="-d extension=/usr/lib/php/20151012/json.so"

Build complete.
Don't forget to run 'make test'.


=====================================================================
PHP         : /usr/bin/php7.0 
PHP_SAPI    : cli
PHP_VERSION : 7.0.30-0ubuntu0.16.04.1
ZEND_VERSION: 3.0.0
PHP_OS      : Linux - Linux ubuntu-xenial 4.4.0-128-generic #154-Ubuntu SMP Fri May 25 14:14:58 UTC 2018 i686
INI actual  : /home/vagrant/mongo-php-driver/tmp-php.ini
More .INIs  :   
CWD         : /home/vagrant/mongo-php-driver
Extra dirs  : 
VALGRIND    : Not used
=====================================================================
Running selected tests.
PASS MongoDB\BSON\Int64 roundtripped through BSON on 32-bit platforms [tests/bson/bson-int64-001.phpt] 
PASS MongoDB\BSON\Int64 wraps 64-bit integers on 32-bit platforms [tests/bson/bson-int64-002.phpt] 
PASS MongoDB\BSON\Int64 encoded as 64-bit integer in BSON [tests/bson/bson-int64-003.phpt] 
PASS MongoDB\BSON\Int64 comparisons [tests/bson/bson-int64-compare-001.phpt] 
PASS MongoDB\BSON\Int64 debug output [tests/bson/bson-int64-debug-001.phpt] 
PASS MongoDB\BSON\Int64 get_properties handler (get_object_vars) [tests/bson/bson-int64-get_properties-001.phpt] 
PASS MongoDB\BSON\Int64 get_properties handler (foreach) [tests/bson/bson-int64-get_properties-002.phpt] 
PASS MongoDB\BSON\Int64::jsonSerialize() return value [tests/bson/bson-int64-jsonserialize-001.phpt] 
PASS MongoDB\BSON\Int64::jsonSerialize() with json_encode() [tests/bson/bson-int64-jsonserialize-002.phpt] 
PASS MongoDB\BSON\Int64 serialization [tests/bson/bson-int64-serialization-001.phpt] 
PASS MongoDB\BSON\Int64 unserialization requires "int" string field [tests/bson/bson-int64-serialization_error-001.phpt] 
PASS MongoDB\BSON\Int64 unserialization requires "int" string field to be valid [tests/bson/bson-int64-serialization_error-002.phpt] 
PASS MongoDB\BSON\Int64::__toString() [tests/bson/bson-int64-tostring-001.phpt] 
PASS MongoDB\BSON\Int64 cannot be extended [tests/bson/bson-int64_error-001.phpt] 
=====================================================================
Number of tests :   14                14
Tests skipped   :    0 (  0.0%) --------
Tests warned    :    0 (  0.0%) (  0.0%)
Tests failed    :    0 (  0.0%) (  0.0%)
Expected fail   :    0 (  0.0%) (  0.0%)
Tests passed    :   14 (100.0%) (100.0%)
---------------------------------------------------------------------
Time taken      :    0 seconds
=====================================================================
```

File: phongo_compat.h [hunk 1]
```diff
@@ -159,29 +159,68 @@
...
+		zval* zchild = NULL;                             \
+		TSRMLS_FETCH();                                  \
+		MAKE_STD_ZVAL(zchild);                           \
+		php_phongo_new_int64(zchild, (_value) TSRMLS_CC); \
+		add_assoc_zval((_zv), (_key), zchild);           \
+	} else {                                             \
+		add_assoc_long((_zv), (_key), (_value));          \
+	}
+#endif /* PHP_VERSION_ID >= 70000 */
+#else  /* SIZEOF_PHONGO_LONG == 8 */
```
Comment (derickr) [inline phongo_compat.h]: The comment here makes it look like this is the end of the "size_long = 8" section, which is not the case, as it's the end of the "size_long = 4" section.

File: src/BSON/Int64.c [hunk 2]
```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright 2014-2017 MongoDB, Inc.
```
Comment (derickr) [inline src/BSON/Int64.c]: It's not 2017.

File: src/BSON/Int64.c [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+/* Initialize the object from a numeric string and return whether it was
+ * successful. An exception will be thrown on error. */
+static bool php_phongo_int64_init_from_string(php_phongo_int64_t* intern, const char* s_integer, phongo_zpp_char_len s_integer_len TSRMLS_DC) /* {{{ */
+{
+	int64_t integer;
+	char*   endptr = NULL;
+
+	/* bson_ascii_strtoll() sets errno if conversion fails. If conversion
+	 * succeeds, we still want to ensure that the entire string was parsed. */
+
```
Comment (derickr) [inline src/BSON/Int64.c]: No newline needed here.

File: src/BSON/Int64.c [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+} /* }}} */
+
+/* Initialize the object from a HashTable and return whether it was successful.
+ * An exception will be thrown on error. */
+static bool php_phongo_int64_init_from_hash(php_phongo_int64_t* intern, HashTable* props TSRMLS_DC) /* {{{ */
+{
+#if PHP_VERSION_ID >= 70000
+	zval* value;
+
+	if ((value = zend_hash_str_find(props, "int", sizeof("int") - 1)) && Z_TYPE_P(value) == IS_STRING) {
```
Comment (derickr) [inline src/BSON/Int64.c]: It's called ``integer`` in variable name and as property of "intern". To be consistent, the property name in the hash should probably be the same, and not just ``int``. I'm not too fussed about this though.

File: src/BSON/Int64.c [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+	char                s_integer[24];
+	int                 s_integer_len;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	intern = Z_INT64_OBJ_P(getThis());
+
+	s_integer_len = snprintf(s_integer, sizeof(s_integer), "%" PRId64, intern->integer);
```
Comment (derickr) [inline src/BSON/Int64.c]: Although I prefer spprintf in general, I guess it makes sense to use a stack buffer here.

File: src/BSON/Int64.c [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+	php_phongo_int64_ce->create_object = php_phongo_int64_create_object;
+	PHONGO_CE_FINAL(php_phongo_int64_ce);
+
+	zend_class_implements(php_phongo_int64_ce TSRMLS_CC, 1, php_phongo_json_serializable_ce);
+	zend_class_implements(php_phongo_int64_ce TSRMLS_CC, 1, php_phongo_type_ce);
+	zend_class_implements(php_phongo_int64_ce TSRMLS_CC, 1, zend_ce_serializable);
+
+	memcpy(&php_phongo_handler_int64, phongo_get_std_object_handlers(), sizeof(zend_object_handlers));
+	php_phongo_handler_int64.compare_objects = php_phongo_int64_compare_objects;
+	php_phongo_handler_int64.get_debug_info  = php_phongo_int64_get_debug_info;
```
Comment (derickr) [inline src/BSON/Int64.c]: Would it make sense to add a cast handler? Like GMP does: https://github.com/php/php-src/blob/master/ext/gmp/gmp.c#L412

File: tests/bson/bson-int64-001.phpt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+--FILE--
+<?php
+
+require_once __DIR__ . '/../utils/tools.php';
+
+/* Omit testing values in the range of 32-bit integers. They will come back from
+ * BSON as a PHP integer type, as Int64 is only used when necessary. */
+$tests = [
+    (object) ['int64' => unserialize('C:18:"MongoDB\BSON\Int64":43:{a:1:{s:3:"int";s:19:"9223372036854775807";}}')],
+    (object) ['int64' => unserialize('C:18:"MongoDB\BSON\Int64":44:{a:1:{s:3:"int";s:20:"-9223372036854775808";}}')],
```
Comment (derickr) [inline tests/bson/bson-int64-001.phpt]: Please add test cases that covers the two "just out of 32-bit range" cases as well.

File: tests/bson/bson-int64-002.phpt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+--FILE--
+<?php
+
+require_once __DIR__ . '/../utils/tools.php';
+
+$tests = [
+    MongoDB\BSON\toPHP(MongoDB\BSON\fromJSON('{"max64": {"$numberLong": "9223372036854775807"}}')),
+    MongoDB\BSON\toPHP(MongoDB\BSON\fromJSON('{"min64": {"$numberLong": "-9223372036854775808"}}')),
+    // Numbers within the range of 32-bit integers will not be wrapped
+    MongoDB\BSON\toPHP(MongoDB\BSON\fromJSON('{"max32": {"$numberLong": "2147483647"}}')),
```
Comment (derickr) [inline tests/bson/bson-int64-002.phpt]: Same here, please add tests just out of the range as well (i.e. 2147483648 and - 2147483649)

File: tests/bson/bson-int64-003.phpt [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+MongoDB\BSON\Int64 encoded as 64-bit integer in BSON
+--FILE--
+<?php
+
+require_once __DIR__ . '/../utils/tools.php';
+
+$tests = [
+    ['int64' => unserialize('C:18:"MongoDB\BSON\Int64":43:{a:1:{s:3:"int";s:19:"9223372036854775807";}}')],
+    ['int64' => unserialize('C:18:"MongoDB\BSON\Int64":44:{a:1:{s:3:"int";s:20:"-9223372036854775808";}}')],
+    ['int64' => unserialize('C:18:"MongoDB\BSON\Int64":24:{a:1:{s:3:"int";s:1:"0";}}')],
```
Comment (derickr) [inline tests/bson/bson-int64-003.phpt]: And here too, please add the just out of range cases.

File: tests/bson/bson-int64-compare-001.phpt [hunk 10]
```diff
@@ -0,0 +1,34 @@
+--TEST--
+MongoDB\BSON\Int64 comparisons
+--FILE--
+<?php
+
+$max = unserialize('C:18:"MongoDB\BSON\Int64":43:{a:1:{s:3:"int";s:19:"9223372036854775807";}}');
+$min = unserialize('C:18:"MongoDB\BSON\Int64":44:{a:1:{s:3:"int";s:20:"-9223372036854775808";}}');
+$zero = unserialize('C:18:"MongoDB\BSON\Int64":24:{a:1:{s:3:"int";s:1:"0";}}');
```
Comment (derickr) [inline tests/bson/bson-int64-compare-001.phpt]: How do comparisons between normal ints and this Int64 class work?

File: phongo_compat.h (hunk 1 above)
Comment (jmikola) [reply phongo_compat.h]: This is no different than `#else /* PHP_VERSION_ID >= 70000 */` used above. IIRC, we discussed this a while back in another PR and the conclusion was that we should stick to the original `#if` condition to match things up. Would you prefer I write this as `SIZEOF_PHONGO_LONG != 8 && SIZEOF_PHONGO_LONG != 4` to be most accurate?

On the same token, I should probably add `SIZEOF_PHONGO_LONG != 8` to the `#endif` two lines below.

File: src/BSON/Int64.c (hunk 4 above)
Comment (jmikola) [reply src/BSON/Int64.c]: Happy to change. I was trying to be consistent with Decimal128, although I'll admit it has always bugged me that it used "dec" instead of "decimal".

File: src/BSON/Int64.c (hunk 5 above)
Comment (jmikola) [reply src/BSON/Int64.c]: Aye, this static buffer approach was done in a few places elsewhere. I see no reason we can't do this any time we're creating a string from a 64-bit integer, which is why I also refactored UTCDateTime in #857. That said, if you have the same concerns about appending a null pointer, would you prefer I use `spprintf` in all of these places instead?

File: src/BSON/Int64.c (hunk 6 above)
Comment (jmikola) [reply src/BSON/Int64.c]: That might alleviate some concerns I had with https://github.com/mongodb/mongo-php-library/pull/542/files/677838f5ca8030942686ec24e1e6b5407ffdeac4..b75386e36c6e824fd045c7ab0bcb5d65efba4b65#r197290561, although I'm not sure it will make much difference in practice.

This object will only be created if the integer's size exceeds the range of PHP's native integer type. In that case, attempting to cast it to `(integer)` is going to certainly result in a loss of data. I think the only safe approach for users is to cast as a string and they can then use that value to construct GMP objects if they want to work with the numbers on a 32-bit platform.

If this class is constructed on a 64-bit integer platform, it's because someone unserialized some data from a 32-bit platform (as I do in tests). That's a very narrow edge case.

File: tests/bson/bson-int64-compare-001.phpt (hunk 10 above)
Comment (jmikola) [reply tests/bson/bson-int64-compare-001.phpt]: I didn't look into that. Do compare handlers facilitate comparisons against scalar types? If so, I'm happy to research that today (any links to prior art would be appreciated), add it, and create some tests.

File: src/BSON/Int64.c (hunk 5 above)
Comment (jmikola) [reply src/BSON/Int64.c]: Note: we agreed to keep use of `snprintf`.

File: src/BSON/Int64.c (hunk 6 above)
Comment (jmikola) [reply src/BSON/Int64.c]: Note: we agreed not to implement a cast handler, since that could result in overflow errors.

File: tests/bson/bson-int64-compare-001.phpt (hunk 10 above)
Comment (jmikola) [reply tests/bson/bson-int64-compare-001.phpt]: I've opened [PHPC-1213](https://jira.mongodb.org/browse/PHPC-1213) and [PHPC-1214](https://jira.mongodb.org/browse/PHPC-1214) to track implementation of compare and do_operation handlers.
